girder_worker/docker/tasks/__init__.py

```diff
@@ -419,34 +461,109 @@ def _docker_run(task, image, pull_image=True, entrypoint=None, container_args=No
-    :param entrypoint: Alternative entrypoint to use when running the container.
-    :type entrypoint: str
-    :param container_args: Arguments to pass to the container.
-    :type container_args: list
-    :param volumes: Volumes to expose to the container.
-    :type volumes: dict
-    :param remove_container: Whether to delete the container after the task is done.
-    :type remove_container: bool
-    :return: Fulfilled result hooks.
-    :rtype: list
-    """
-    return _docker_run(
-        task, image, pull_image, entrypoint, container_args, volumes,
-        remove_container, **kwargs)
+# @app.task(base=DockerTask, bind=True)
+# def docker_run(task, image, pull_image=True, entrypoint=None, container_args=None,
+#                volumes=None, remove_container=True, **kwargs):
+#     """
+#     This task runs a docker container. For details on how to use this task, see the
+#     :ref:`docker-run` guide.
+
+#     :param task: The bound task reference.
+#     :type task: :py:class:`girder_worker.task.Task`
+#     :param image: The docker image identifier.
+#     :type image: str
+#     :param pull_image: Whether to explicitly pull the image prior to running the container.
+#     :type pull_image: bool
+#     :param entrypoint: Alternative entrypoint to use when running the container.
+#     :type entrypoint: str
+#     :param container_args: Arguments to pass to the container.
+#     :type container_args: list
+#     :param volumes: Volumes to expose to the container.
+#     :type volumes: dict
+#     :param remove_container: Whether to delete the container after the task is done.
+#     :type remove_container: bool
+#     :return: Fulfilled result hooks.
+#     :rtype: list
+#     """
+#     return _docker_run(
+#         task, image, pull_image, entrypoint, container_args, volumes,
+#         remove_container, **kwargs)
+
+@app.task(base=SingularityTask, bind=True)
+def singularity_run(task, image, *args, container_args=None, bind_paths=None, **kwargs):
+    return task(image,*args,container_args,bind_paths,**kwargs)
+
+#This function is used to check whether we need to switch to singularity or not.
+def use_singularity():
+    runtime = os.environ.get('RUNTIME')
+    if runtime == 'SINGULARITY':
+        return True
+    if runtime == 'DOCKER':
+        return False
+    try:
+        #Check whether we are connected to a docker socket.
+        with socket.socket(socket.AF_UNIX, socket.SOCK_STREAM) as s:
+            return s.connect_ex('/var/run/docker.sock') != 0
+    except socket.error:
+        return True
+    
+def _generate_slurm_script(image, container_args, bind_paths, qos, cpus, gpus, memory, other_slurm_options):
+    # Construct the bind option for the Singularity command
+    bind_option = ','.join([f"{host}:{container}" for host, container in bind_paths.items()])
+
+    # Construct the Singularity command
+    singularity_command = f'singularity exec --bind {bind_option} docker://{image} {" ".join(container_args)}'
```

Comment:
The join is unsafe -- if the container args contain spaces or quotes or shell characters, odd things will happen.